src/bids_validator/context.py

```diff
@@ -129,14 +130,44 @@ def dataset_description(self) -> Namespace:
     @cached_property
     def modalities(self) -> list[str]:
         """List of modalities found in the dataset."""
-        ...
-        return []
+        result = set()
+
+        modalities = self.schema.rules.modalities
+        for datatype in self.datatypes:
+            for mod_name, mod_dtypes in modalities.items():
+                if datatype in mod_dtypes.datatypes:
+                    result.add(mod_name)
+
+        return list(result)
 
     @cached_property
     def datatypes(self) -> list[str]:
         """List of datatypes found in the dataset."""
-        ...
-        return []
+        datatypes = self.schema.objects.datatypes
+        result = find_datatypes(self.tree, datatypes)
+
+        return list(result)
```

Comment:
No longer needs to be so spread out.

```suggestion
        return list(find_datatypes(self.tree, self.schema.objects.datatypes))
```